### PR TITLE
[#133][WIP][Netty] Basic framework

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?

Realize the basic framework of netty replace grpc.

### Why are the changes needed?
Rewrite grpc interface to interface implemented with netty.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

UT.

### CODE REVIEW

**Comment by leixm** [comment]
to-do list:
1. Remove redundant code
2. Implement other interfaces except sendShuffleData
3. The client supports retry logic
4. The client supports asynchronous callbacks instead of CountDownLatch
5. Add more UTs, the old integration tests need to support grpc and netty tests at the same time

**Comment by leixm** [comment]
to-do list
1. Remove redundant code & fix UT
2. Implement other interfaces except sendShuffleData
3. The client supports retry logic
4. The client supports asynchronous callbacks instead of CountDownLatch
5. Add more UTs, the old integration tests need to support grpc and netty tests at the same time

**Comment by jerqi** [comment]
Could we split this pr into more smaill prs?

**Comment by leixm** [comment]
> Could we split this pr into more smaill prs?
OK.

File: common/src/main/java/org/apache/uniffle/common/ClientType.java [hunk 1]
```diff
@@ -18,5 +18,6 @@
 package org.apache.uniffle.common;
 
 public enum ClientType {
-  GRPC
+  GRPC,
+  GRPC_NETTY
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/ClientType.java:22]: I will split the `ClientType` and `ShuffleServerInfo` to part-1.

**Comment by leixm** [comment]
Points 3 and 4 will pull out two more PRs, and I will complete the development on the basis of #711.

**Comment by leixm** [comment]
> to-do list
> 
> 1. Remove redundant code & fix UT
> 2. Implement other interfaces except sendShuffleData
> 3. The client supports retry logic
> 4. The client supports asynchronous callbacks instead of CountDownLatch
> 5. Add more UTs, the old integration tests need to support grpc and netty tests at the same time

Points 3 and 4 will pull out two more PRs, and I will complete the development on the basis of #711.

**Comment by jerqi** [comment]
Could we split this into 4 - 5 parts?
1. A streamServer
2. Netty Utils
3. Encoder and Decoder
4. ShuffleNettyGrpcClient
5. Server's concrete service

**Comment by leixm** [comment]
> Could we split this into 4 - 5 parts?
> 
> 1. A streamServer
> 2. Netty Utils
> 3. Encoder and Decoder
> 4. ShuffleNettyGrpcClient
> 5. Server's concrete service
> 6. Off heap Memory Manager


Does point 5 mean that all interfaces are implemented in ShuffleServer? There are already other PRs for point 6, and this has been separated out when it was split before.

**Comment by jerqi** [comment]
> > Could we split this into 4 - 5 parts?
> > 
> > 1. A streamServer
> > 2. Netty Utils
> > 3. Encoder and Decoder
> > 4. ShuffleNettyGrpcClient
> > 5. Server's concrete service
> > 6. Off heap Memory Manager
> 
> Does point 5 mean that all interfaces are implemented in ShuffleServer? There are already other PRs for point 6, and this has been separated out when it was split before.

1. Yes.
2. ok for me.

**Comment by jerqi** [comment]
There are other prs to finish the feature, we close it.

**Pull request closed without merging**